### PR TITLE
schedule:  one operator only occupy one limit

### PR DESCRIPTION
**Pull request by bufferflies**
Signed-off-by: tongjian <1045931706@qq.com>

### What problem does this PR solve?
fix: #3807 
	 #3778

### What is changed and how it works?
operator controller's count will occupy the highest priority OpKind even if the operator has many OpKind. 

the relation between limit and scheduler is blow:
image

red mark need to be changed 
### Check List

Tests

- Unit test

Code changes

Side effects

Related changes

### Release note

```release-note
Reduce the conflict due to multiple scheduler running in same time
```

### CODE REVIEW

File: server/schedule/operator/operator.go [hunk 1]
```diff
@@ -125,6 +125,18 @@ func (o *Operator) Kind() OpKind {
 	return o.kind
 }
 
+// SchedulerKind return the highest OpKind even if the operator has many OpKind
+// fix #3778
+func (o *Operator) SchedulerKind() OpKind {
+	priority := []OpKind{OpMerge, OpRange, OpReplica, OpHotRegion, OpRegion, OpLeader}
```
Comment (HunDunDM) [inline server/schedule/operator/operator.go:131]: Is `OpRange` really needed here?

**Review by HunDunDM — COMMENTED** [review]
`OperatorCount` can `return oc.counts[mask]` directly  now?

File: server/schedule/operator/operator.go [hunk 2]
```diff
@@ -125,6 +125,18 @@ func (o *Operator) Kind() OpKind {
 	return o.kind
 }
 
+// SchedulerKind return the highest OpKind even if the operator has many OpKind
+// fix #3778
+func (o *Operator) SchedulerKind() OpKind {
```
Comment (HunDunDM) [inline server/schedule/operator/operator.go:130]: We can rearrange OpKind and use the LowBit function to get it directly.
`LowBit(x) = x&(-x)`

File: server/schedule/operator/operator.go [hunk 3]
```diff
@@ -125,6 +125,16 @@ func (o *Operator) Kind() OpKind {
...
+	// LowBit ref: https://zh.wikipedia.org/wiki/%E6%A0%91%E7%8A%B6%E6%95%B0%E7%BB%84
```
Comment (HunDunDM) [inline server/schedule/operator/operator.go]: This is a Chinese page. You can use `https://en.wikipedia.org/wiki/Find_first_set`.

File: server/schedule/operator_controller.go [hunk 4]
```diff
@@ -769,7 +769,7 @@ func (oc *OperatorController) updateCounts(operators map[uint64]*operator.Operat
 		delete(oc.counts, k)
 	}
 	for _, op := range operators {
-		oc.counts[op.Kind()]++
+		oc.counts[op.SchedulerKind()]++
```
Comment (HunDunDM) [inline server/schedule/operator_controller.go:772]: Maybe we can implement a new `OperatorCount` function without changing the existing function?

File: server/schedule/operator/kind.go [hunk 5]
```diff
@@ -24,22 +24,23 @@ type OpKind uint32
...
-	// Initiated by merge checker or merge scheduler. Note that it may not include region merge.
-	OpMerge
-	// Initiated by range scheduler.
-	OpRange
+	// Include peer addition or removal. This means that this operator may take a long time.
+	OpRegion
+	// Include leader transfer.
+	OpLeader
+	// Initiated by admin.
+	OpAdmin
```
Comment (nolouch) [inline server/schedule/operator/kind.go:43]: should Admin with higher priority?

Comment (HunDunDM) [reply server/schedule/operator/kind.go:43]: LGTM, `OpAdmin` should not occupy the limit.

Comment (rleungx) [reply server/schedule/operator/kind.go:43]: It will make `op show admin` cannot be used.

Comment (HunDunDM) [reply server/schedule/operator/kind.go:43]: > It will make `op show admin` cannot be used.

No, it uses `GetOperatorsOfKind`.

Comment (rleungx) [reply server/schedule/operator/kind.go:43]: Ok

File: server/schedule/operator/operator.go (hunk 2 above)
Comment (rleungx) [reply server/schedule/operator/operator.go:130]: Perhaps, we can find a better name.

**Review by HunDunDM — APPROVED** [review]

**Review by lhy1024 — COMMENTED** [review]
Do we need additional test?

**Review by nolouch — APPROVED** [review]
lgtm

**Pull request merged**